
# Analysis
## Somatic Interactions
```{r,message=FALSE,results='markup'}
try(somaticInteractions(maf = dataForReport[["maf"]], top = 10,genes=dataForReport[["interestedGenes"]], pvalue = c(0.05, 0.1)))
```

<!-- ## Detecting cancer driver genes based on positional clustering -->
```{r,results='hide'}
if (dataForReport[["genome"]] %in% c("hg18","hg19","hg38")) {
  cat("## Detecting cancer driver genes based on positional clustering\n")
  mafOncodrive = try(oncodrive(maf = dataForReport[["maf"]], AACol = dataForReport[["AACol"]], minMut = 5, pvalMethod = 'zscore'))
}
```

```{r}
if (class(mafOncodrive)=="try-error") { #error in oncodrive, usually because of not enough genes with many mutations, can change minMut to smaller value.
  
} else {
  if (dataForReport[["genome"]] %in% c("hg18","hg19","hg38")) {
  makeDataTable(mafOncodrive[,c("Hugo_Symbol",setdiff(colnames(mafOncodrive),colnames(dataForReport[["maf"]]@gene.summary))),with = FALSE])
  
  fdrCut=max(sort(mafOncodrive$fdr)[5]+0.00000001,0.05) #make sure fdrCut can include at least 5 points for plot
  
  if (nrow(mafOncodrive)>200) {
    plotInd=1:200
    warning(paste0("Too many Oncodrive clusters: ",nrow(mafOncodrive),". Only top 200 were plotted. See table above for more details."))
    print(paste0("Too many Oncodrive clusters: ",nrow(mafOncodrive),". Only top 200 were plotted. See table above for more details."))
  } else {
    plotInd=1:nrow(mafOncodrive)
  }
  if (0) { #Too slow now. Not to do plot at this time
    plotOncodrive(res = mafOncodrive[plotInd,], fdrCutOff = fdrCut, useFraction = TRUE)
  }
  }
}

```

<!-- ## Summarizing pfam domains -->
```{r}
if (dataForReport[["genome"]] %in% c("hg18","hg19","hg38")) {
  cat("## Summarizing pfam domains\n")
  mafObj=dataForReport[["maf"]]
  #remove not Standard AaChange. Something like "wholegene" will report error. Add ; behind it to aviod it
  mafObj@data[[dataForReport[["AACol"]]]]=gsub("^[A-Za-z]+$","\\1;",dataForReport[["maf"]]@data[[dataForReport[["AACol"]]]])
  mafPfamDomains= pfamDomains(maf = mafObj, AACol = dataForReport[["AACol"]], top = 10)
}
```


<!-- ## Drug-Gene Interactions -->
```{r,results='markup'}
if (dataForReport[["genome"]] %in% c("hg18","hg19","hg38")) {
  cat("## Drug-Gene Interactions\n")
  mafDrugInteractions = drugInteractions(maf = dataForReport[["maf"]], fontSize = 0.75)
}
```



<!-- ## Oncogenic Signaling Pathways -->
```{r,results='markup'}
if (dataForReport[["genome"]] %in% c("hg18","hg19","hg38")) {
  cat("## Oncogenic Signaling Pathways\n")
  mafOncogenicPathways = capture.output(OncogenicPathways(maf = dataForReport[["maf"]]))
#plot(mafOncogenicPathways)
  mafOncogenicPathways=fread(text=mafOncogenicPathways)
  if (nrow(mafOncogenicPathways)>0) {
    dataForPlot=head(mafOncogenicPathways,3)
    for (i in 1:nrow(dataForPlot)) {
      (PlotOncogenicPathways(maf = dataForReport[["maf"]], pathways = dataForPlot$Pathway[i]))
    }
  }

}
```


<!-- ## Tumor heterogeneity and MATH scores -->
<!-- Top 5 most mutated samples were used -->
```{r}
if (dataForReport[["performVaf"]]) {
  cat("## Tumor heterogeneity and MATH scores")
  mafInferHeterogeneity = inferHeterogeneity(maf = dataForReport[["maf"]], vafCol = dataForReport[["vafCol"]])
  
  #makeDataTable(mafInferHeterogeneity$clusterMeans)
  out=list(makeDataTable(mafInferHeterogeneity$clusterMeans))
  print(tagList(out))
  
  plotClusters(clusters = mafInferHeterogeneity)
}

```

<!-- ```{r} -->
<!-- mafInferHeterogeneity = inferHeterogeneity(maf = dataForReport[["maf"]], vafCol = 'i_TumorVAF_WU') -->
<!-- gvisTable(mafInferHeterogeneity$clusterMeans,options=tableVisOptions) -->
<!-- plotClusters(clusters = mafInferHeterogeneity) -->

<!-- ``` -->

## Mutational Signatures

```{r MutationalSignatures,results='markup',eval=FALSE}
if (dataForReport[["genome"]]=="mm10") {
  genomePackage=paste0("BSgenome.Mmusculus.UCSC.",dataForReport[["genome"]])
} else {
  genomePackage=paste0("BSgenome.Hsapiens.UCSC.",dataForReport[["genome"]])
}

library(genomePackage, quietly = TRUE,character.only=TRUE)
#mafTnm = trinucleotideMatrix(maf = dataForReport[["maf"]], prefix = 'chr', add = TRUE, ref_genome =dataForReport[["genomePackage"]])

#check "chr" issue. Please note we changed seqlevelsStyle of maf file. May influence other analysis in future. So need to use another variable
genomePkgStype=suppressWarnings(GenomeInfoDb::seqlevelsStyle(BSgenome::getBSgenome(genome =genomePackage)))
mafObj=dataForReport[["maf"]]
mafObj@data$Chromosome=as.character(mafObj@data$Chromosome)
mafObj@maf.silent$Chromosome=as.character(mafObj@maf.silent$Chromosome)
GenomeInfoDb::seqlevelsStyle(mafObj@data$Chromosome)=genomePkgStype
GenomeInfoDb::seqlevelsStyle(mafObj@maf.silent$Chromosome)=genomePkgStype

mafTnm = trinucleotideMatrix(maf = mafObj,ref_genome =genomePackage)
try(plotApobecDiff(tnm = mafTnm, maf = mafObj))

library('NMF')
##old vertsion
#mafTnmSign=try(extractSignatures(mat = mafTnm, plotBestFitRes = FALSE))
#try(plotSignatures(mafTnmSign, title_size = 1 ,yaxisLim=NA))

#with R4.0
mafTnmSign = estimateSignatures(mat = mafTnm, nTry = 6)
#mafTnmSign = estimateSignatures(mat = mafTnm, nTry = 6, pConstant = 0.1, plotBestFitRes = FALSE, parallel = 2)
#plotCophenetic(res = mafTnmSign)
mafTnmSign = extractSignatures(mat = mafTnm, n = 5)

#Compate against original 30 signatures 
mafTnmSign.og30.cosm = compareSignatures(nmfRes = mafTnmSign, sig_db = "legacy")
#Compate against updated version3 60 signatures 
mafTnmSign.v3.cosm = compareSignatures(nmfRes = mafTnmSign, sig_db = "SBS")

library('pheatmap')
pheatmap::pheatmap(mat = mafTnmSign.v3.cosm$cosine_similarities, cluster_rows = FALSE, main = "cosine similarity against validated signatures")

## ---- fig.width=6, fig.height=4, fig.align='center', eval = T-----------------
maftools::plotSignatures(nmfRes = mafTnmSign, title_size = 1.2, sig_db = "SBS")

#plot contribution to each sample
library(reshape2)
library(ggplot2)

dataForPlot=melt(mafTnmSign$contributions)
colnames(dataForPlot)=c("Signature","Sample","Contribution")
ggplot(dataForPlot,aes(x=Sample,y=Contribution,fill=Signature))+geom_bar(stat = "identity")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))



```


<!-- ## dndscv -->
```{r dndscv}
if (!is.null(dataForReport[["dndscv.refdb"]])) {
  cat("## dndscv\n")
  
#  library(dndscv)
  mafDatSnp=rbind(dataForReport[["maf"]]@data,dataForReport[["maf"]]@maf.silent)
  #mafDatSnp=mafDatSnp[which(mafDatSnp$Variant_Type=="SNP"),]
  
  #remove In_Frame_Del or In_Frame_Ins in DnDs analysis if they were not considered as vc_nonSyn
  if (!("In_Frame_Del" %in% dataForReport$vc_nonSyn)) {
    mafDatSnp=mafDatSnp[which(! mafDatSnp$Variant_Classification %in% c("In_Frame_Del")),]
  }
  if (!("In_Frame_Ins" %in% dataForReport$vc_nonSyn)) {
    mafDatSnp=mafDatSnp[which(! mafDatSnp$Variant_Classification %in% c("In_Frame_Ins")),]
  }
  
  dataForModel=mafDatSnp[,.(Tumor_Sample_Barcode ,Chromosome,Start_Position,Reference_Allele,Tumor_Seq_Allele2)]
  dataForModel$Chromosome=as.character(dataForModel$Chromosome)

  #check chr style in data
  if (dataForReport[["dndscv.refdb"]] == "hg19") {
     data("refcds_hg19", package = "dndscv")
  } else {
     load(dataForReport[["dndscv.refdb"]])
  }
  genomePkgStype=GenomeInfoDb::seqlevelsStyle(sapply(RefCDS,function(x) x$chr))
  GenomeInfoDb::seqlevelsStyle(dataForModel$Chromosome)=genomePkgStype
  #colnames(dataForModel)<-c("sampleID","chr","pos","ref","alt")

  dndsout=try(dndscv::dndscv(dataForModel,cv=NULL,refdb=dataForReport[["dndscv.refdb"]],use_indel_sites=FALSE))
  #print(names(dndsout))
  
  if (class(dndsout)!="try-error") {
    makeDataTable(dndsout$sel_cv[1:500,])
  }
  
#  kable(dndsout$globaldnds,row.names=FALSE)
}

if (!is.null(dataForReport[["dndscv.refdb"]])) {  #have to be in a seprate section to make sure makeDataTable(dndsout$sel_cv[1:500,]) can display
  if (class(dndsout)!="try-error") {
    kable(dndsout$globaldnds,row.names=FALSE)
  }
}
```





